### ISSUE TITLE
Conflicting udev rules causing failure to boot with root on multi-device btrfs

### ISSUE THREAD

**Issue by greghensley** [rockstor/rockstor-core#1329]
It is well known that Rockstor currently does not support booting to a multi-device btrfs root (hence, the official [mirrored boot device tutorial](http://rockstor.com/docs/mdraid-mirror/boot_drive_howto.html)), because the btrfs device scan is not being performed early in the boot process.

However, the root cause of the issue appears to be `/etc/udev/rules.d/64-btrfs.rules`.  This udev rule file is installed by the post-install script of the `rockstor` RPM package (copied from `/opt/rockstor/conf/64-btrfs.rules`), and does nothing more than create the `/dev/btrfs-control` device file.

First, it appears that the version of systemd, shipped with CentOS 7, is already capable of creating the btrfs-control file, automatically, so the rule file is redundant.

Second, and the cause of the failure to boot, this file shadows systemd's own btrfs rules: `/usr/lib/udev/rules.d/64-btrfs.rules`.  Systemd's rule file correctly runs `btrfs ready $devnode` for any device, found by udev, for which `blkid` identifies as a btrfs filesystem.  But, since there is already a rule file with a matching name in `/etc/udev/rules.d`, systemd's rule file is never loaded.  This can be verified by booting with the `rdudevdebug` kernel parameter.

Simply removing the Rockstor-provided `/etc/udev/rules.d/64-btrfs.conf` and regenerating the initramfs was enough to successfully boot to a 2-disk btrfs mirror.

**Comment by phillxnet** [comment]
@greghensley This is a nice find, and thank for contributing your findings; could you clarify if in your successful trials you still had a separate non btrfs /boot as per our upstream anaconda installer inherited limitation. I'm thinking of the following forum thread investigations:-
https://forum.rockstor.com/t/install-to-full-disk-btrfs/1170/8
and @Spectre694 's find in that thread of grubby's bug:-
https://bugzilla.redhat.com/show_bug.cgi?id=864198

**Comment by greghensley** [comment]
I'm using a separate boot partition, along with an EFI boot partition.  My drive setup looks like this:

```
sde           8:64   0 202.7G  0 disk  
├─sde1        8:65   0   200M  0 part  
│ └─md126     9:126  0   200M  0 raid1 /boot/efi (vfat)
├─sde2        8:66   0   500M  0 part  
│ └─md127     9:127  0   500M  0 raid1 /boot (ext2)
├─sde3        8:67   0    32G  0 part  /mnt2/centos_rockstor
└─sde4        8:68   0     8G  0 part  [SWAP]
sdf           8:80   0 202.7G  0 disk  
├─sdf1        8:81   0   200M  0 part  
│ └─md126     9:126  0   200M  0 raid1 /boot/efi
├─sdf2        8:82   0   500M  0 part  
│ └─md127     9:127  0   500M  0 raid1 /boot
├─sdf3        8:83   0    32G  0 part  /mnt2/centos_rockstor
└─sdf4        8:84   0     8G  0 part  [SWAP]
```

Where sde3 and sdf3 form the btrfs volume (label: centos_rockstor) containing the root and home subvols.  I wasn't feeling adventurous enough to put swap on a loopback-mounted file in btrfs, and I had to partition for UEFI, anyway, so I just kept /boot separate for simplicity.

**Comment by greghensley** [comment]
Oh, and I forgot to mention in the original report that my install is non-standard: I started with a vanilla CentOS 7, with root and home on a single-device btrfs on top of an md-raid1 of sd[ef]3.  The rockstor packages were installed over that after adding the rockstor, epel, and elrepo repositories.

Also, this issue is more generalized than just booting -- it applies to the availability of any multi-device btrfs volume during the boot process.  I originally installed the system using the officially-recommended md-raid setup, so I only noticed there was a problem after I went about creating some LXC containers in a subvol on my main pool (a btrfs raid1, not shown in previous drive layout) and realized they weren't starting automatically on boot because the subvol wasn't mounted yet.  Adding the lxc container subvol to my fstab revealed the problem: same error as when trying to boot to a multi-device root (I don't remember the exact message, something to the effect of "can't find the btrfs superblock on <dev>").

Once I realized that the rockstor-bootstrap service was handling the mounting of rockstor-managed volumes, a quick systemd target to force the lxc service to wait for rockstor-bootstrap resolved the issue of auto-starting containers.  But, being me, I still didn't like not being able to add unmanaged subvols to my fstab and having to use custom systemd rules.  After tracking down the cause of the missing ```btrfs device ready/scan``` calls, everything fell in to place and I converted the md-raid1 to a btrfs raid1.

**Comment by phillxnet** [comment]
@greghensley Great info here, thanks for the additional background and discovery exposition. Think this one need the eyes of @schakrava and @MFlyer. Pretty sure the separate /boot is actually a requirement for the time being at least, as referenced in the prior forum thread.

Is you system actually booting in UEFI mode? My understanding was that uefi bios was not able to decode mdraid and so wouldn't see your md126 device in order to access it's required FAT format.

Thanks.

**Comment by greghensley** [comment]
> My understanding was that uefi bios was not able to decode mdraid and so wouldn't see your md126 device in order to access it's required FAT format.

You're right: the EFI boot partition has to be FAT and the UEFI firmware does not understand md-raid. And I've read in multiple places that md-raid can't be used in that context. But I think that's just a case of misinformation being copied around that originated with someone's misunderstanding. The key is in the [md-raid superblock](https://raid.wiki.kernel.org/index.php/RAID_superblock_formats#Sub-versions_of_the_version-1_superblock). By using version 1.0 superblocks, the md-raid superblock is put at the end of the member volumes, instead of the beginning, so the filesystem starts at the beginning of the volume and the UEFI firmware sees two identical FAT filesystems.

Anaconda's GUI lets you put `/boot/efi` on an md-raid, but the result is broken -- I assume it's using v1.2 metadata and I've no idea what it puts it the UEFI boot entries. I just installed CentOS with EFI boot on a basic partition, then manually converted to a raid1 and fixed up the boot entries (one for each mirror member).

**Comment by greghensley** [comment]
Since this issue hasn't seen any movement, I went looking for the RPM spec file to submit a patch... but I don't see it in any of Rockstor's public repos.  Am I just not seeing it (inside an archive, maybe?), or does it only live on a build server somewhere?  Any objection to the packaging scripts being included in [rockstor-core](https://github.com/rockstor/rockstor-core)?

**Comment by haoto** [comment]
@greghensley Thanks so much for the find! I have the same setup and have been plagued by this bug for more than a year. Before this I had to use yum versionlock kernel-ml to prevent kernel updates when I'm not around.

@phillxnet Any chance we can get this fixed in Rockstor? Thanks!

**Comment by Luke-Nukem** [comment]
Can confirm this solves boot issues and prevents the need to the work-around in #1980. PR will be incoming to remove this 7 year old rule.

**Comment by Hooverdan96** [comment]
@phillxnet while you suggested a fresh issue/PR for this (in closed/unmerged PR #1984), should we keep this open instead? Or start a new one?

**Comment by phillxnet** [comment]
@Hooverdan96 Thanks for the triage work here.
Re:

> should we keep this open instead? Or start a new one?

Given we have changed base OS from CentOS to openSUSE Leap & TW and now resource kiwi-ng as our installer, this is now a legacy issue. Plus the referenced mdraid doc approach referenced is also legacy now so I've opened the following doc issue for that howto's removal:

https://github.com/rockstor/rockstor-doc/issues/466

For the sake of tying up loose ends here @greghensley re:

> I went looking for the RPM spec file to submit a patch... but I don't see it in any of Rockstor's public repos.

We have for some time now (since 4.5.7-0) had the following dedicated repo, with matching master/testing versions that mirror the master/testing branches in rockstor-core and rockstor-jslibs, and carry the same GitHub release/tag versioning/timing.

Give the indicated udev rules have also now since been removed. I'll close this as legacy related only.
